Network/Socket/Types.hsc

```diff
@@ -97,15 +98,22 @@ mkSocket fd = do
     void $ mkWeakIORef ref $ close s
     return s
 
+invalidateSocket ::
+      Socket
+   -> (CInt -> IO a)
+   -> (CInt -> IO a)
+   -> IO a
+invalidateSocket (Socket ref) errorAction normalAction = do
+    oldfd <- atomicModifyIORef' ref $ \cur -> (-1, cur)
+    if oldfd == -1 then errorAction oldfd else normalAction oldfd
+
 -----------------------------------------------------------------------------
 
 -- | Close the socket. Sending data to or receiving data from closed socket
 --   may lead to undefined behaviour.
```

Comment:
As it turns out, we no longer have that problem. When a socket is closed, any *later* I/O operations will encounter an invalid `fd` and raise an error (bad file descriptor or similar).  The only issue is that in multi-thread programs if one thread is doing I/O and another is closing the socket, *then* the thread that got the `fd` just before it was closed, might be using a stale, still valid value.

This is in a semantic improvement.  We now have safety for sequential operations.  And only risk undefined behaviour in the unlikely case that a thread is closing a socket while another thread is using it.